src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -199,7 +199,17 @@ def close(self) -> None:
         self.__closed = True
 
     def join(self, timeout: Optional[float] = None) -> None:
-        self.__flush()
+        start = time.time()
+        while self.__batch:
+            remaining = timeout - (time.time() - start) if timeout is not None else None
+            if remaining is not None and remaining <= 0:
+                logger.warning(f"Timed out with {len(self.__batch)} messages left in batch")
+                break
+            try:
+                self.__flush()
+            except MessageRejected:
+                pass
```

Comment:
I wonder if we could ignore this part entirely. 
If we received 100 messages here and we still did not start processing them, if we throw them away in this case, they will not be committed. 
So whether join is called upon termination or rebalancing. If we do not flush, those messages will not be processeed thus not committed and, correctly, will be consumed again by the consumer that will get the same partition assigned.

Up to you.

Reply:
Actually thinking over it, this will not work. 
If you keep calling submit on the next step but never poll, the next step will never complete its work. 

I think you should just throw the batch away without committing and let the new consumer consume them again.

Reply:
> Actually thinking over it, this will not work. If you keep calling submit on the next step but never poll, the next step will never complete its work.
> 
> I think you should just throw the batch away without committing and let the new consumer consume them again.

So basically we'd always timeout (if we had a timeout to begin with) since submit would never succeed right? But actually since we don't even have a timeout, this would go on forever

Reply:
exactly